src/services/codefixes/addOptionalPropertyUndefined.ts

```diff
@@ -0,0 +1,121 @@
+/* @internal */
+namespace ts.codefix {
+    const addOptionalPropertyUndefined = "addOptionalPropertyUndefined";
+
+    const errorCodes = [
+        Diagnostics.Type_0_is_not_assignable_to_type_1_with_exactOptionalPropertyTypes_Colon_true_Consider_adding_undefined_to_the_types_of_the_target_s_properties.code,
+        Diagnostics.Argument_of_type_0_is_not_assignable_to_parameter_of_type_1_with_exactOptionalPropertyTypes_Colon_true_Consider_adding_undefined_to_the_types_of_the_target_s_properties.code
+    ];
+
+    registerCodeFix({
+        errorCodes,
+        getCodeActions(context) {
+            const typeChecker = context.program.getTypeChecker();
+            const toAdd = getPropertiesToAdd(context.sourceFile, context.span.start, typeChecker);
+            if (!toAdd.length) {
+                return undefined;
+            }
+            const changes = textChanges.ChangeTracker.with(context, t => addUndefinedToOptionalProperty(t, toAdd));
+            return [createCodeFixAction(addOptionalPropertyUndefined, changes, Diagnostics.Add_undefined_to_optional_property_type, addOptionalPropertyUndefined, Diagnostics.Add_undefined_to_all_optional_properties)];
+        },
+        fixIds: [addOptionalPropertyUndefined],
+        getAllCodeActions: context => {
+            const { program } = context;
+            const checker = program.getTypeChecker();
+            const seen = new Map<string, true>();
+            return createCombinedCodeActions(textChanges.ChangeTracker.with(context, changes => {
+                eachDiagnostic(context, errorCodes, diag => {
+                    const toAdd = getPropertiesToAdd(diag.file, diag.start, checker);
+                    if (!toAdd.length) {
+                        return;
+                    }
+                    let untouched = true;
+                    for (const add of toAdd) {
+                        if (!addToSeen(seen, add.id + "")) {
+                            untouched = false;
+                        }
+                    }
+                    if (untouched) {addUndefinedToOptionalProperty(changes, toAdd);}
+                });
+            }));
+        },
+    });
+
+    function getPropertiesToAdd(file: SourceFile, pos: number, checker: TypeChecker): Symbol[] {
+        const sourceTarget = getSourceTarget(getErrorNode(file, pos), checker);
+        if (!sourceTarget) {
+            return [];
+        }
+        const { source: sourceNode, target: targetNode } = sourceTarget;
+        const target = checker.getTypeAtLocation(targetNode);
+        if (target.symbol?.declarations?.some(d => getSourceFileOfNode(d).fileName.match(/(node_modules|^lib\.)/))) {

```

Comment:
Pretty sure we have other code actions that know not to touch `node_modules` and `lib`. Is there some way you can look into reusing that logic here?

Reply:
There’s `program.isSourceFileFromExternalLibrary` which includes “files found by searching under node_modules” (not sure exactly what the difference is between this and “files in node_modules”).

I would also argue that maybe we shouldn’t run on `.d.ts` files at all since they’re likely to be autogenerated and overwritten during some other compilation or codegen step.

Reply:
In a word: no. There are one-offs scattered throughout the compiler+services but nothing in codefixes. I'll take the code from TODO comment scraping and move it to services/utilities.ts, so at least it'll be a *little* shared.

Reply:
Later: github doesn't always load comments, so I missed @andrewbranch 's and was replying to @DanielRosenwasser 's.